transform.ml

```diff
@@ -1,205 +1,124 @@
-  let rec loop pc acc =
-    if pc = Array.length instrs then acc
-    else match[@warning "-4"] instrs.(pc-1), instrs.(pc) with
-    | Goto _, Label _
-    | Branch _, Label _ ->
-      loop (pc+1) acc
-    | _, Label l ->
-      let edit = (pc, 0, [| Goto l |]) in
-      loop (pc+1) (edit :: acc)
-    | _, _ ->
-      loop (pc+1) acc
-  in
-  let edits = loop 1 [] in
-  fst (Edit.subst_many instrs edits)
+(* combining transformations *)
+let (&&&) t2 t1 inp =
+  match t1 inp with
+  | None -> t2 inp
+  | Some out ->
+    begin match t2 out with
+      | None -> Some out
+      | Some out' -> Some out'
+    end
 
+let try_opt t inp =
+  match t inp with
+  | None -> inp
+  | Some out -> out
 
-(* Hoisting assignments "x <- exp" as far up the callgraph as possible.
- *
- * Since we are not in SSA moving assignments is only possible (without further
- * analysis) if the assignments dominates all its uses. Otherwise we might
- * accidentally capture uses on some traces.
- *
- * The condition for a move to be valid is that the move target dominates the
- * move origin (ie. we are moving upwards) and is dominated by all reaching
- * defs (ie. we are not moving above our dependencies).
- *
- * We only look at our own use-def chain. Thus the transformation renames the
- * variable to avoid overriding unrelated uses of the same name.
- *)
-let hoist_assignment (func : afunction) : afunction =
-  let version = Instr.active_version func in
-  let instrs = version.instrs in
-  let reaching = Analysis.reaching instrs in
-  let uses = Analysis.used instrs in
-  let dominates = Analysis.dominates instrs in
-  let dominates_all_uses pc =
-    Analysis.PcSet.for_all (fun use -> dominates pc use) (uses pc) in
-  let rec find_possible_move pc =
-    if pc = Array.length instrs then None
-    else
-      let pc' = pc + 1 in
-      match[@warning "-4"] instrs.(pc) with
-      | Assign (x, exp) ->
-        if not (dominates_all_uses pc) then find_possible_move pc'
-        else
-          let reaching_defs = reaching pc in
-          let valid_move candidate =
-            let dominate_me = Analysis.PcSet.for_all (fun pc -> dominates pc candidate) in
-            dominates candidate pc && dominate_me reaching_defs in
+(* generalizing transformations *)
+let as_opt_function (transformation : transform_instructions) : opt_function =
+  fun func ->
+    let version = Instr.active_version func in
+    let inp = Analysis.as_analysis_input func version in
+    match transformation inp with
+    | None -> None
+    | Some instrs ->
+      let res = {
+        label = version.label;
+        instrs;
+        annotations = None } in
+      Some (Instr.replace_active_version func res)
 
-          begin match Analysis.find_first instrs valid_move with
-          | exception Not_found -> find_possible_move pc'
-          | pc' -> Some (pc, pc')
-          end
+let as_opt_program (transform : opt_function) : opt_prog =
+  let transform instrs =
```

Comment:
Could you rename this? It shadows the `transform` parameter, and I was confused why on line 56, `transform prog.main` was returning two values.